### PR TITLE
Database version 3

### PR DESCRIPTION
**Pull request by peterwilsoncc**
This will help speed up preflight filters.

`args(50)` portion of the index is pretty much a random number.

### CODE REVIEW

**Comment by roborourke** [comment]
Code looks good, I'm just wondering about the implications of the args index and whether we need to change that number or not.

**Comment by rmccue** [comment]
Could we index by just hook instead? 🤔 wondering how much improvement we gain from this

**Comment by peterwilsoncc** [comment]
@rmccue Do you mean instead of `hook_args` or all the other indexes too?

In order of frequency as `WHERE` clauses across the runner and plugin (with preflight filters), frequency of use is:

* `nextrun` and `status`
* `site`
* `id` (few occurrences in code but used by runner for each job run)
* `hook`
* `args`

**Comment by rmccue** [comment]
Just `hook_args`. If we just have a `hook` index, that'll get used, and then in-memory filtering can take place for the args; hooks are unique enough that I think this should have fine performance (as evidenced by us not even having these indexes right now). My concern is that `hook_args` is going to be a biiiiiiiiig index and likely has too high cardinality to actually be useful.

File: inc/namespace.php [hunk 1]
```diff
@@ -82,7 +82,10 @@ function create_tables() {
 		`schedule` varchar(255) DEFAULT NULL,
 
 		PRIMARY KEY (`id`),
-		KEY `status` (`status`)
+		KEY `status` (`status`),
+		KEY `site` (`site`),
+		KEY `hook_args` (`hook`, `args`(50)),
```
Comment (roborourke) [inline inc/namespace.php]: As per @rmccue's comment lets just index `hook` on its own

File: inc/upgrade/namespace.php [hunk 2]
```diff
@@ -62,3 +66,19 @@ function upgrade_database_2() {
...
+ * Upgrade Cavalcade database tables to version 3.
+ *
+ * Add indexes required for pre-flight filters.
+ */
+function upgrade_database_3() {
+	global $wpdb;
+
+	$query = "ALTER TABLE `{$wpdb->base_prefix}cavalcade_jobs`
+			  ADD INDEX `site` (`site`),
+			  ADD INDEX `hook_args` (`hook`, `args`(50)),
```
Comment (roborourke) [inline inc/upgrade/namespace.php]: Just index `hook` as above

File: inc/namespace.php (hunk 1 above)
Comment (peterwilsoncc) [reply inc/namespace.php]: `args` removed from index in 3227c6b797f2661a9959a2e0c5076434885c85ef.

I've got the _allow edits from maintainers_ option checked on this PR so you should be able to push to my repo if needs be, presuming GH's permissions model is entirely bug free.

**Pull request merged**